add docs for recorded commands (#56944)
sourcegraph/sourcegraph: doc/admin/repo/recording.md

# Configuring command recording

> Command recording allows site admins to view all git operations executed on a repository. 

Reviewer: Would be great to have a screenshot here. It would explain a lot of the other things here.

When enabled, Sourcegraph will record metadata about all git commands run on a repository in Redis, including:

- Command executed (with sensitive information redacted)
- Execution time  
- Duration of execution
- Success state
- Output

> This provides visibility into git operations performed by Sourcegraph on a repository, which can be useful for debugging and monitoring.

Reviewer: Does it also include Perforce? Or not at all? I think we should explicitly call this out.

To enable command recording:

> - Go to Site Admin > Site Configuration

Reviewer: I think we should link to the doc page, see: https://sourcegraph.sourcegraph.com/search?q=context:sourcegraph+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+file:doc+content:%22site_config.md%29%22&patternType=standard&sm=0&groupBy=path

Author: Done.

> 1. Go to Site Admin > Site Configuration

Reviewer 2: ```suggestion
- Go to Site Admin > Site Configuration
```

Reviewer 2: Can we please use bullets instead? 🙏

> 2. Add a `gitRecorder` object to the configuration object

Reviewer 2: ```suggestion
- Add a `gitRecorder` object to the configuration object
```

> ```json
> "gitRecorder": {
>     // the amount of commands to record per repo  
>     "size": 30,
>   
>     // repositories to record commands. This can either be a wildcard '*' 
>     // to record commands for all repositories or a list of repositories
>     "repos": ["*"],
>   
>     // git commands to exclude from recording. We exclude the 
>     // commands below by default.
>     "ignoredGitCommands": [
>       "show",
>       "rev-parse",
>       "log",
>       "diff",
>       "ls-tree"
>     ]
>   }
> ```

Reviewer: Can we make this a valid JSON object so it renders without errors?

```suggestion
```jsonc
{
  // [...]

  "gitRecorder": {
      // the amount of commands to record per repo  
      "size": 30,
    
      // repositories to record commands. This can either be a wildcard '*' 
      // to record commands for all repositories or a list of repositories
      "repos": ["*"],
    
      // git commands to exclude from recording. We exclude the 
      // commands below by default.
      "ignoredGitCommands": [
        "show",
        "rev-parse",
        "log",
        "diff",
        "ls-tree"
      ]
   }
}
```
```

I think you can also use `jsonc` here. (See: https://sourcegraph.sourcegraph.com/search?q=context:sourcegraph+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+file:doc+jsonc&patternType=standard&sm=0&groupBy=path)

Once enabled, site admins can view recorded commands for a repository via the repository's settings page in the Site Admin UI.

> Recorded commands include information like start time, duration, exit status, command executed, directory, and output. Sensitive information like usernames, passwords, tokens are automatically redacted from the command and output.

Reviewer 2: ```suggestion
Recorded commands include information like start time, duration, exit status, command executed, directory, and output. Sensitive information like usernames, passwords, and tokens are automatically redacted from the command and output.
```

Command recording provides visibility into Sourcegraph's interactions with repositories without requiring modifications to Sourcegraph's core Git operations.

> ### Potential Risks

Reviewer 2: ```suggestion
### Potential risks
```
We don't title case the headings if they are not product names.

> Depending on the number of repositories and size of the recording set, enabling command recording could increase disk usage in Redis. 

Reviewer: I'm pretty sure it will increase disk usage, not just "could" :) I think the thing to call out here is that it could significantly change disk usage, which you do in the 2nd paragraph. So I'd combine them.

Author: Gotcha.

> Depending on the number of repositories and size of the recording set, enabling command recording could result in increased disk usage in Redis. 

Reviewer 2: ```suggestion
Depending on the number of repositories and size of the recording set, enabling command recording could increase disk usage in Redis. 
```

Since recorded commands are stored in Redis, setting the `size` to a very large number or enabling recording on many repositories could cause the Redis database to fill up quickly.

> When Redis is full, it may start evicting data, impacting other parts of Sourcegraph that rely on Redis. Sourcegraph may experience degraded performance or instability.

Reviewer: I'd add something like "Depending on your configuration"

> When Redis is full, it may start evicting data which can impact other parts of Sourcegraph that rely on Redis. Sourcegraph may experience degraded performance or instability.

Reviewer 2: ```suggestion
When Redis is full, it may start evicting data, impacting other parts of Sourcegraph that rely on Redis. Sourcegraph may experience degraded performance or instability.
```
